src/mqtt_client.rs

```diff
@@ -59,6 +59,27 @@ impl<E> From<E> for Error<E> {
     }
 }
 
+mod sm {
+
+    use smlang::statemachine;
+
+    statemachine! {
+        transitions: {
+            *Restart + Update = ConnectTransport,
```

Comment:
I would differentiate the two `Update` events into `GotSocket` and `SentConnect`.